### DISCUSSION TITLE
[Help]: Is their a way to run a function after the animation is finished

### DISCUSSION THREAD

**Discussion by michaelkremenetsky** [davidjerleke/embla-carousel#814] [Help] [answered]
### Summary

So I have a carousel that I control with the arrow keys and updates the query params in the URL on slide change using the nextJS replace function. Issue is that changing the query params causes a bit of stutter at the end of the animation. I fixed this when I was using nuka-carousel by using the `afterSlide` function but switched to embla-carousel as there were issues with that lib. Is there a way to run the function after the animation is done similar to `afterSlide` in nuka-carousel (https://commerce.nearform.com/open-source/nuka-carousel/api#afterslide)?

Snippet of how I am doing it currently:
```tsx
  const { replace, asPath } = useRouter();
  const pathWithoutQuery = asPath.split("?")[0];

  const debouncedReplace = useRef(
    debounce((debouncedItems, endSlideIndex) => {
      replace(
        `${pathWithoutQuery}?item=` + debouncedItems[endSlideIndex]!.id,
        undefined,
        { shallow: true, scroll: false },
      );
    }, 500),
  );

  useEffect(() => {
    return () => {
      debouncedReplace.current.cancel();
    };
  }, []);

  const [emblaRef, emblaApi] = useEmblaCarousel({
    startIndex: initialIndex,
    containScroll: "keepSnaps",
    watchDrag: false,
    duration: fullscreen ? 0 : 25,
  });

  const slideChange = () => {
    const newIndex = emblaApi?.selectedScrollSnap();

    if (newIndex) {
      if (newIndex == newBufferedItems.length - 1) {
        fetchNextPage();
      }

      if (newBufferedItems[newIndex] != undefined) {
        markRead(newBufferedItems[newIndex]);
      }

      debouncedReplace.current(newBufferedItems, newIndex);
    }
  };

  useHotkeys("LeftArrow", () => {
    if (emblaApi) {
      emblaApi.scrollPrev();
      slideChange();
    }
  });
  useHotkeys("RightArrow", () => {
    if (emblaApi) {
      emblaApi?.scrollNext();
      slideChange();
    }
  });
```

[output.webm](https://github.com/davidjerleke/embla-carousel/assets/28833521/febe4cf1-079b-4ede-b261-1e0ae6058547)

### If applicable, which variants of Embla Carousel are relevant to this question?

- [X] embla-carousel **(Core)**
- [X] embla-carousel-react
- [ ] embla-carousel-vue
- [ ] embla-carousel-svelte
- [ ] embla-carousel-autoplay
- [ ] embla-carousel-auto-scroll
- [ ] embla-carousel-solid
- [ ] embla-carousel-auto-height
- [ ] embla-carousel-class-names
- [ ] embla-carousel-docs **(Documentation)**
- [ ] embla-carousel-docs **(Generator)**

**Comment by davidjerleke** [comment]
@michaelkremenetsky thanks for your question. Do you want your function to trigger after the animation ends or as soon as a new active slide has been selected?

**Reply by michaelkremenetsky** [reply]
Thanks for the extremely speedy response! I wanna do both in this case. I wanna change the query param after the animation ends as that's causing the stutter. But the fetchNextPage() and markRead() can run as soon as it's selected.

**Reply by davidjerleke** [reply, marked as answer]
@michaelkremenetsky I see. I think you're looking for these events:

- [`select`](https://www.embla-carousel.com/api/events/#select)
- [`settle`](https://www.embla-carousel.com/api/events/#settle)

And here is an [example](https://www.embla-carousel.com/api/events/#adding-event-listeners) of how to make use of these events in the docs. Remember to switch to the `React` tab for a relevant code example.

**Reply by michaelkremenetsky** [reply]
Thank you, `settle` worked! I had I tried `select` earlier but somehow missed `settle`.
